### PR TITLE
valueChanges firing multiple

### PR DESCRIPTION
**Pull request by Blackbaud-TrevorBurch**

### CODE REVIEW

File: src/app/public/modules/colorpicker/colorpicker.component.spec.ts [hunk 1]
```diff
@@ -186,443 +211,640 @@ describe('Colorpicker Component', () => {
...
+      tick();
+      fixture.detectChanges();
+      tick();
+      expect(nativeElement.querySelectorAll('.sky-colorpicker-reset-button').length).toEqual(1);
+    }));
+
+    it('should only emit the form control valueChanged event once per change', async(() => {
+      fixture.detectChanges();
+      let count = 0;
+      component.colorForm.valueChanges.subscribe(() => {
```
Comment (Blackbaud-AlexKingman) [inline src/app/public/modules/colorpicker/colorpicker.component.spec.ts]: Since this is an async test, have you double checked to make sure its waiting until completion to do all `valueChanges` events are fired? Also, why not just use a `spy` here?

Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/colorpicker/colorpicker.component.spec.ts]: I didn't use a spy due to the fact that I need to wait for all async things to be done. A spy might deliver a false result on the initial expectation. I need some way to asynchronously ensure that everything has completed before marking the tests as passed. 

I did verify that if I remove my changes this test fails due to `valueChanges` firing more times.

Comment (Blackbaud-AlexKingman) [reply src/app/public/modules/colorpicker/colorpicker.component.spec.ts]: Ok good! I had this same issue with an async test last week. I solved it with a setTimeout, but maybe this is the better approach.

Well if you've confirmed it, then I suppose this looks good to me. 👍

File: src/app/visual/colorpicker/colorpicker-visual.component.html [hunk 2]
```diff
@@ -1,21 +1,21 @@
 <div id="screenshot-colorpicker">
   <sky-colorpicker #colorPickerExample1>
-    <input
+  <input
```
Comment (Blackbaud-SteveBrush) [inline src/app/visual/colorpicker/colorpicker-visual.component.html]: I believe you accidentally removed the tabs in this file.

File: src/app/public/modules/colorpicker/colorpicker-input.directive.ts [hunk 3]
```diff
@@ -95,7 +101,8 @@ export class SkyColorpickerInputDirective
     private elementRef: ElementRef,
     private renderer: Renderer,
     private service: SkyColorpickerService,
-    private resourcesService: SkyLibResourcesService
+    private resourcesService: SkyLibResourcesService,
+    private inj: Injector
```
Comment (Blackbaud-SteveBrush) [inline src/app/public/modules/colorpicker/colorpicker-input.directive.ts]: We should avoid abbreviated variable names, so there's less room for confusion later on.
`private injector: Injector`

File: src/app/public/modules/colorpicker/colorpicker-input.directive.ts [hunk 4]
```diff
@@ -192,6 +198,10 @@ export class SkyColorpickerInputDirective
         this.skyColorpickerInput.initialColor = value;
       }
       this.skyColorpickerInput.lastAppliedColor = value;
+      let control: FormControl = (<NgControl>this.inj.get(NgControl)).control as FormControl;
+      if (control) {
```
Comment (Blackbaud-SteveBrush) [inline src/app/public/modules/colorpicker/colorpicker-input.directive.ts]: Could you use something simpler to get the FormControl, like this?
```
constructor(
  @Optional() private formControl: NgControl
) { }

if (this.formControl) {
  // ...
}
```

File: src/app/public/modules/colorpicker/colorpicker.component.spec.ts [hunk 5]
```diff
@@ -186,443 +211,640 @@ describe('Colorpicker Component', () => {
...
-    fixture.detectChanges();
-    tick();
-    expect(nativeElement.querySelectorAll('.sky-colorpicker-reset-button').length).toEqual(0);
-    colorpickerComponent.showResetButton = true;
-    tick();
-    fixture.detectChanges();
-    tick();
-    expect(nativeElement.querySelectorAll('.sky-colorpicker-reset-button').length).toEqual(1);
-  }));
+  fdescribe('reactive configuration', () => {
```
Comment (Blackbaud-SteveBrush) [inline src/app/public/modules/colorpicker/colorpicker.component.spec.ts]: `fdescribe`

File: src/app/public/modules/colorpicker/colorpicker.component.spec.ts [hunk 6]
```diff
@@ -186,443 +211,640 @@ describe('Colorpicker Component', () => {
...
-    fixture.detectChanges();
-    directiveInstance.ngOnInit();
-    tick();
-    fixture.detectChanges();
-    tick();
-    openColorpicker(nativeElement, fixture);
-    fixture.detectChanges();
-    expect(nativeElement.querySelectorAll('.sky-colorpicker-hidden').length).toEqual(1);
-  }));
+    it('should populate correct information if model is given', fakeAsync(() => {
```
Comment (Blackbaud-SteveBrush) [inline src/app/public/modules/colorpicker/colorpicker.component.spec.ts]: The number of changes in this spec is a tad unsettling. Can you give me some color around the changes? If it's purely for cleanup, maybe we could do that in a different pull request?

I would prefer we keep the changes specific to your feature/bug fix, as it helps me be more confident that nothing else was inadvertently broken or changed.

File: src/app/public/modules/colorpicker/colorpicker.component.spec.ts (hunk 1 above)
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/colorpicker/colorpicker.component.spec.ts]: I'm inclined to agree with Alex. I also think we should have an `expect` in here somewhere.

File: src/app/public/modules/colorpicker/colorpicker-input.directive.ts (hunk 4 above)
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/colorpicker/colorpicker-input.directive.ts]: I'll try. I don't believe so based on what I was finding online as I researched but I might be wrong

File: src/app/public/modules/colorpicker/colorpicker.component.spec.ts (hunk 6 above)
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/colorpicker/colorpicker.component.spec.ts]: No. The changes are due to a bad assumption around how things were working. The `valueChanges` was firing on every mouse drag and updating the model before anything was saved. This was allowing earlier access to the new value than it should have been. The unit tests were taking advantage of this and I had to fix the tests to now work with what was decided to be the correct behavior. Nothing about the tests spirit changed but I had to make some changes to get them to work with the intended behavior.

File: src/app/public/modules/colorpicker/colorpicker-input.directive.ts (hunk 3 above)
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/colorpicker/colorpicker-input.directive.ts]: Will do

File: src/app/visual/colorpicker/colorpicker-visual.component.html (hunk 2 above)
Comment (Blackbaud-TrevorBurch) [reply src/app/visual/colorpicker/colorpicker-visual.component.html]: I'll give it a look

File: src/app/public/modules/colorpicker/colorpicker.component.spec.ts (hunk 5 above)
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/colorpicker/colorpicker.component.spec.ts]: Oops. Will fix

File: src/app/public/modules/colorpicker/colorpicker-input.directive.ts (hunk 4 above)
Comment (Blackbaud-SteveBrush) [reply src/app/public/modules/colorpicker/colorpicker-input.directive.ts]: Good deal. Thanks for checking.

File: src/app/visual/colorpicker/colorpicker-visual.component.html (hunk 2 above)
Comment (Blackbaud-TrevorBurch) [reply src/app/visual/colorpicker/colorpicker-visual.component.html]: This has been fixed

File: src/app/public/modules/colorpicker/colorpicker.component.spec.ts (hunk 1 above)
Comment (Blackbaud-TrevorBurch) [reply src/app/public/modules/colorpicker/colorpicker.component.spec.ts]: Done
